presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java

```diff
@@ -0,0 +1,165 @@
+            if (stripe.getDataLength() + stripe.getFooterLength() + stripe.getIndexLength() > MAX_CACHE_LENGTH) {
+                return dataSource;
+            }
+        }
+        return new CachingOrcDataSource(dataSource, stripes);
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return dataSource.getReadTimeNanos();
+    }
+
+    @Override
+    public long getSize()
+    {
+        return dataSource.getSize();
+    }
+
+    @VisibleForTesting
+    void readCacheAt(long offset)
+            throws IOException
+    {
+        // Assumption: stripes are always read in order
+        // Assumption: bytes that are not part of any stripe is never read
+        long startOffset;
+        long endOffset;
+        while (true) {
+            verify(stripeIterator.hasNext());
+            StripeInformation firstStripe = stripeIterator.next();
+            long firstStripeLength = firstStripe.getIndexLength() + firstStripe.getDataLength() + firstStripe.getFooterLength();
+            verify(firstStripeLength <= MAX_CACHE_LENGTH);
+            startOffset = firstStripe.getOffset();
+            endOffset = firstStripe.getOffset() + firstStripeLength;
+            if (endOffset > offset) {
+                verify(startOffset <= offset);
+                break;
+            }
+        }
+        while (stripeIterator.hasNext()) {
+            StripeInformation nextStripe = stripeIterator.next();
+            long nextStripeLength = nextStripe.getIndexLength() + nextStripe.getDataLength() + nextStripe.getFooterLength();
+            long newEndOffset = nextStripe.getOffset() + nextStripeLength;
+            if (newEndOffset - startOffset > MAX_CACHE_LENGTH) {
+                stripeIterator.previous();
+                break;
+            }
+            endOffset = newEndOffset;
+        }
+        cacheLength = Ints.checkedCast(endOffset - startOffset);
+        expandCacheIfNeeded(cacheLength);
+        dataSource.readFully(startOffset, cache, 0, cacheLength);
+        cachePosition = startOffset;
+    }
+
+    private void expandCacheIfNeeded(int requiredLength)
+    {
+        if (cache == null || cache.length < requiredLength) {
+            cache = new byte[requiredLength];
+        }
+    }
+
+    @Override
+    public void readFully(long position, byte[] buffer)
+            throws IOException
+    {
+        readFully(position, buffer, 0, buffer.length);
+    }
+
+    @Override
+    public void readFully(long position, byte[] buffer, int bufferOffset, int length)
+            throws IOException
+    {
+        if (position < cachePosition) {
+            throw new IllegalArgumentException(String.format("read request (offset %d length %d) is before cache (offset %d length %d)", position, length, cachePosition, cacheLength));
+        }
+        if (cache == null || position > cachePosition + cacheLength) {
+            readCacheAt(position);
+        }
+        if (position + length > cachePosition + cacheLength) {
```

Comment:
This check happens after you read. Is this actually verifying an invariant (i.e. it is an assertion)? If so, use `verify()`